hashicorp/dev-portal
# Add Checkbox

## 🔗 Relevant links



- Preview urls 🔎 
  - [`CheckboxField`](https://dev-portal-git-ambadd-checkbox-hashicorp.vercel.app/swingset/components/checkboxfield)
  - [`CheckboxGroup`](https://dev-portal-git-ambadd-checkbox-hashicorp.vercel.app/swingset/components/checkboxgroup)
- [Asana task](https://app.asana.com/0/1202097197789424/1202816749416416/f) 🎟️

## 🗒️ What



- Adds new `src/components/form` directory
  - Adds a basic `Form` component, not used anywhere but there and ready for when it will be used
- Adds several subdirectories in `src/components/form`:
  - `components`: for generic atoms that can be used to compose fields/groups
  - `base-controls`: for the base elements used in fields/groups, like `<input>`s with specific `type`s set
  - `field-controls`: for piecing together atoms and base controls
  - `group-controls`: for piecing together several fields, and some usage of atoms
- Adds `Errors` atom: renders an icon and a list of error messages. Can be used at the field or group level, and its styles never change between the two uses.
- Adds `HelperText` atom: renders text that describes a field or group. Its styles never change between the two uses. In fields, it should be referenced as the element that describes the base control via `aria-describedby` on the control and `id` on the `HelperText`.
- Adds `Label` atom: renders a `<label>` element and has `font-weight` customizations for use at the group level when a group does not have a legend.
- Adds `Legend` atom: renders a `<legend>` with reusable (and customizable) styles.
- Adds `CheckboxControl`: an `<input>` of type `checkbox` and has quite a bit of custom styling that's good to reuse
- Adds `CheckboxField` and Swingset page: uses a few atoms and `CheckboxControl` to create a field that can include a label, helper text, and error messages. Handles associating `CheckboxControl` and `Label` with `useId`.
- Adds `CheckboxGroup` and Swingset page: uses `CheckboxField` and some atoms to show at the group level. Does not expose `children` as prop because it handles several things for us with less code and higher quality markup:
  - Not showing helper text at the `CheckboxField` level when `layout=horizontal`
  - Setting the `Label` font weight based on whether or not there is a `Legend`
  - Setting the `name` prop for all rendered `CheckboxField`
  - Not showing errors at the `CheckboxField` level
- Updates Swingset `componentsRoot` blob to enable nesting `docs.mdx` files in subdirectories of `src/components`
- Upgrades the `@hashicorp/design-system-tokens` package so the new form-related tokens are accessible
  - Removes reference to deprecated `--token-color-neutral-border-strong` in `src/swingset-docs/color.mdx`
  - Replaces reference to deprecated `--token-color-neutral-border-strong` with `--token-color-border-strong ` in `src/views/profile/bookmarks-view/components/empty-state/empty-state.module.css`

## 📸 Design Screenshots



- [HDS Checkbox Figma file](https://www.figma.com/file/noyY6dUMDYjmySpHcMjhkN/HDS-Product---Components?node-id=9120%3A23131)
- [HDS Checkbox documentation page](https://design-system-components-hashicorp.vercel.app/components/form/checkbox)

## 🧪 Testing



- Go to the [`CheckboxField` Swingset page](https://dev-portal-git-ambadd-checkbox-hashicorp.vercel.app/swingset/components/checkboxfield)
- Make sure elements in the examples are styled correctly according to docs
- Make sure the elements behave as you expect
- Go to the [`CheckboxGroup` Swingset page](https://dev-portal-git-ambadd-checkbox-hashicorp.vercel.app/swingset/components/checkboxgroup)
- Make sure elements in the examples are styled correctly according to docs
- Make sure the elements behave as you expect

## 💭 Anything else?



Future component ideas:

- `FieldSet`, which would handle setting up the `grid` approach used in `CheckboxField` for example
  - Maybe it would have some validation of `children` and/or automatically place certain children in the grid based on the order of the `children`? This would reduce some CSS responsibility of placing elements in their appropriate `grid-area`.
